src/CalcManager/CEngine/History.cpp

```diff
@@ -449,30 +436,25 @@ void CHistoryCollector::UpdateHistoryExpression(uint32_t radix, int32_t precisio
         return;
     }
 
-    unsigned int size;
-    IFT(m_spTokens->GetSize(&size));
-
-    for (unsigned int i = 0; i < size; ++i)
+    for (auto token : *m_spTokens)
     {
-        std::pair<std::wstring, int> token;
-        IFT(m_spTokens->GetAt(i, &token));
         int commandPosition = token.second;
         if (commandPosition != -1)
         {
-            std::shared_ptr<IExpressionCommand> expCommand;
-            IFT(m_spCommands->GetAt(commandPosition, &expCommand));
+            const std::shared_ptr<IExpressionCommand>& expCommand = m_spCommands->at(commandPosition);
+
             if (expCommand != nullptr && CalculationManager::CommandType::OperandCommand == expCommand->GetCommandType())
             {
-                std::shared_ptr<COpndCommand> opndCommand = std::static_pointer_cast<COpndCommand>(expCommand);
+                const std::shared_ptr<COpndCommand>& opndCommand = std::static_pointer_cast<COpndCommand>(expCommand);
                 if (opndCommand != nullptr)
                 {
                     token.first = opndCommand->GetString(radix, precision);
-                    IFT(m_spTokens->SetAt(i, token));
```

Comment:
Nice catch.  Instead of setting a new copy of the token, would it make sense to have `token` declared as `auto&`.  Then any changes should be updated in place.

Reply:
Yes, it would work!